### PR TITLE
fix DBImpl::NewInternalIterator super-version leak on failure

### LINKED ISSUE #2955: Superversion leaks in DBImpl::NewInternalIterator on error return.
### Expected behavior
Superversion is registered for cleanup at all times or released immediately

### Actual behavior
Superversion is only registered for cleanup for happy path. Otherwise the pointer is left dangling as NewErrorInternalIterator() does not care about it.

### Steps to reproduce the behavior
See https://github.com/facebook/rocksdb/blob/master/db/db_impl.cc#L911

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Summary:
Close #2955 

Test Plan:
Add a new test and run the test with ASAN.

### CODE REVIEW

File: db/db_impl.cc [hunk 1]
```diff
@@ -907,6 +907,15 @@ InternalIterator* DBImpl::NewInternalIterator(
     internal_iter->RegisterCleanup(CleanupIteratorState, cleanup, nullptr);
 
     return internal_iter;
+  } else {
+    // cleanup super version.
```
Comment (yiwu-arbug) [inline db/db_impl.cc]: @ajkr I'm not familiar with the logic here. Should I call CleanupIteratorState instead?

Comment (ajkr) [reply db/db_impl.cc]: I think in other cases where we hold the SuperVersion for a very short time, like in `Get()`, we don't cleanup everything, so there's a small chance obsoleted files will still exist until the next background job happens.

You should use `ReturnAndCleanupSuperVersion`, though.

Comment (yiwu-arbug) [reply db/db_impl.cc]: The SuperVersion is obtained via `cfd->GetReferencedSuperVersion()` which is already returned to the ThreadLocalPtr. Here we only need the cleanup part, so `ReturnAndCleanupSuperVersion` does not work.

**Comment by yiwu-arbug** [comment]
@ajkr Updated. The added test shows more issue, which is good.

**Comment by yiwu-arbug** [comment]
@ajkr ping

File: db/db_impl.cc [hunk 2]
```diff
@@ -1820,21 +1823,23 @@ SuperVersion* DBImpl::GetAndRefSuperVersion(uint32_t column_family_id) {
...
+      InstrumentedMutexLock l(&mutex_);
+      sv->Cleanup();
+    }
+    delete sv;
+    RecordTick(stats_, NUMBER_SUPERVERSION_CLEANUPS);
+  }
+  RecordTick(stats_, NUMBER_SUPERVERSION_RELEASES);
+}
+
 void DBImpl::ReturnAndCleanupSuperVersion(ColumnFamilyData* cfd,
```
Comment (ajkr) [inline db/db_impl.cc]: thanks for splitting this function. Ideally we'd just use `GetReferencedSuperVersion` everywhere and not need this alternative way to return superversion, right? Maybe we can do it in tech debt week :).

File: db/db_impl.cc [hunk 3]
```diff
@@ -1820,21 +1823,23 @@ SuperVersion* DBImpl::GetAndRefSuperVersion(uint32_t column_family_id) {
...
+  // Release SuperVersion
+  if (sv->Unref()) {
+    {
+      InstrumentedMutexLock l(&mutex_);
+      sv->Cleanup();
+    }
+    delete sv;
+    RecordTick(stats_, NUMBER_SUPERVERSION_CLEANUPS);
+  }
+  RecordTick(stats_, NUMBER_SUPERVERSION_RELEASES);
```
Comment (ajkr) [inline db/db_impl.cc]: idk the meaning of this stat, but it seems inconsistently updated depending how the user accesses superversion. maybe we should just deprecate these during tech debt week

**Pull request closed without merging**